### PR TITLE
MoE Refactor

### PR DESCRIPTION
**Pull request by ludziej**
# Description

Refactor of TC and EC classes
Now, each MoE model consists of 3 parts:
- Gating [TokenGating, ExpertGating], that inherit from MoEGating. This calculates topk indices and values for both types of routing.
- Routing [TokenRouting, ExpertRouting], having two main functions, extract and merge, doing respectively first and second routing parts. Both models use the same configuration (if possible), for example, can be defined using granularity/expasion_rate. 
- inner expert: ExpertFF, ExpertGated, and ExpertLinear, that can be passed to both types of routing independently. Specifically, ExpertGated can implement SwiGLU and GeGLU activations, and ExpertFF supports non-gated activations like GeLU and ReLU. ExpertLinear is just one linear layer. The type of inner expert can be specified using "moe_inner_expert" parameter and the type of its activation by "activation_type". For backward compatibility and brevity, if you specify "moe_inner_expert" as "swi_glu" or "geglu", it will return the proper config that is, ExpertGated with respectively Silu and Gelu activations. This allows for a few new configs to be runnable, for example, if you want you can run ExpertGated with "relu" or "tanh" activations if you want. 

Other functionalities:
- You can still run old models with suffix "_old"
- instead of using "with measure_time("name")" you can now also use function decorator "time_measured("name")" that works the same way. Also, you can wrap any nn.Module with "self.measure(module,name, active=True)", that will be turned off if active is false. This makes the MoE code more readable, at least for me.

All of these functionalities are tested in unit tests for backward compatibility (that is, new Token Choice returns the same gradients as old Token Choice etc), and previous unit tests are rewritten for new code. 

This PR has been split and does not contain double routing. 

I'm going to run experiments to be sure that it works & then I'm going to finish the description and make this non-draft.

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

**Review by Krystian-Krol — COMMENTED** [review]
Jak to skończysz to chętnie przejrzę docelowo, póki co zerknąłem na szybko pobieżnie i wygląda spoko ale takie nity

File: research/conditional/moe_layers/expert_choice.py [hunk 1]
```diff
@@ -205,65 +43,45 @@ def __init__(
...
         self.softmax_over = softmax_over
-        self.extract_chosen_tokens = (
-            self.extract_chosen_tokens_onehot
+        self.extract = (
+            self.extract_bmm
+            if use_torch_bmm
+            else self.extract_einsum
             if one_hot_impl
-            else self.extract_chosen_tokens_select
+            else self.extract_index_select
```
[the remark this answers is not in the document]
Comment (ludziej) [reply research/conditional/moe_layers/expert_choice.py:74]: Right, gonna change.

File: research/conditional/moe_layers/expert_choice.py [hunk 2]
```diff
@@ -363,42 +155,47 @@ def extract_with_linear_bmm(
...
+    # postprocess implementations
+
+    @time_measured("postprocess_einsum")
+    def merge_einsum(
+        self, x, batch_size, topk, seq_len, topk_values, topk_indices, one_hot
+    ):
+        topk //= seq_len
+        x = x.reshape(self.n_experts, topk, seq_len, self.doutput)
+        return einsum(
+            "n_exp topk seq_len dmodel, n_exp topk seq_len, n_exp topk seq_len batch_size "
```
[the remark this answers is not in the document]
Comment (ludziej) [reply research/conditional/moe_layers/expert_choice.py:184]: thanks, forgot about this!

File: research/conditional/moe_layers/load_balancing_loss.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    Calculates the load balancing loss for the token choice layer.
+
+    :param str alpha: aux loss weigth parameter
+    :param torch.Tensor softmax_per_token: tensor of shape (tokens, n_experts)
+    :param torch.Tensor tokens_in_each_expert: tensor of shape (n_experts)
+    """
+    n_tokens, n_experts = softmax_per_token.shape
+    assert n_experts == n_tokens_in_each_expert.shape[0]
+
+    per_expert_softmax_sum = torch.mean(softmax_per_token, dim=0)
```
[the remark this answers is not in the document]
Comment (ludziej) [reply research/conditional/moe_layers/load_balancing_loss.py:21]: No clue, this is just copied from the previous implementation. Since it's a refactor I think it should be left as is (but maybe investigate it later?).

**Comment by ludziej** [comment]
Code-wise it's finished, so you can review it fully. Other things like DoubleRouting and Switch with topk>1 are split into other PRs (that will be merged after this one). I'm gonna remove the draft status after running some exps.

**Comment by ludziej** [comment]
Applied requested changes.

File: research/conditional/moe_layers/expert_choice.py [hunk 4]
```diff
@@ -291,64 +104,39 @@ def forward(self, x: torch.Tensor):
...
-                one_hot,
-            )
-        with measure_time(self, "reshape"):
-            x = x.reshape((self.n_experts, topk, self.dmodel))
+    # extract implementations
+
+    @time_measured("extract_einsum")
+    def extract_einsum(self, x: torch.Tensor, topk, topk_indices: torch.Tensor):
+        batch_size, _, _ = x.shape
+        one_hot = F.one_hot(topk_indices, num_classes=batch_size).type(x.dtype)
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:124]: nie działa z `group_by_batch` jeśli dobrze widzę, bo `topk_indices` ma wtedy wartości z zakresu do `batch_size*seq_len`

File: research/conditional/moe_layers/expert_choice.py [hunk 5]
```diff
@@ -291,64 +104,39 @@ def forward(self, x: torch.Tensor):
...
+        # one_hot is (n_experts, topk, seq_len, batch_size)
+        x = einsum(
+            "batch_size seq_len dmodel, n_exp topk seq_len batch_size "
+            "-> n_exp topk seq_len dmodel",
+            x,
+            one_hot,
+        )
+        x = x.reshape((self.n_experts, topk, self.dmodel))
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:132]: I'm not a fan of this two `topk` meaning different value

File: research/conditional/moe_layers/expert_choice.py [hunk 6]
```diff
@@ -291,64 +104,39 @@ def forward(self, x: torch.Tensor):
...
-            return x, one_hot
-
-    def extract_with_linear_bmm(
-        self, x: torch.Tensor, topk_indices: torch.Tensor, batch_size, weight
-    ):
+    def extract_bmm(self, x: torch.Tensor, topk, topk_indices: torch.Tensor):
+        batch_size, _, _ = x.shape
         with measure_time(self, "one_hot_instanciate"):
             one_hot = F.one_hot(topk_indices, num_classes=batch_size).type(x.dtype)
         n_exp, topk, seq_len, _ = one_hot.shape
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:139]: Incompatible with `group_by_batch`

File: research/conditional/moe_layers/expert_choice.py [hunk 7]
```diff
@@ -363,42 +151,47 @@ def extract_with_linear_bmm(
...
-                one_hot,
-                weight,
-            )
+    def extract_index_select(self, x: torch.Tensor, topk, topk_indices: torch.Tensor):
+        batch_size, _, _ = x.shape
+        # flatten x s. t. first dimension is tokens instead of batch_size x seq_len
+        with measure_time(self, "first_flatten"):
+            x = x.flatten(start_dim=0, end_dim=1)
+        with measure_time(self, "index_select"):
+            x = torch.index_select(x, dim=0, index=topk_indices.flatten())
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:166]: this is only for `group_by_batch` hence I'd rename this function to reflect that somehow

File: research/conditional/moe_layers/expert_choice.py [hunk 8]
```diff
@@ -363,42 +151,47 @@ def extract_with_linear_bmm(
...
+        return x, None
+
+    # postprocess implementations
 
-    def gating_postprocess_bmm(self, x, topk_values, one_hot, weight):
-        n_exp, exp_size, dmodel = weight.shape
-        seq_len, batch_size, _ = one_hot.shape
+    @time_measured("postprocess_einsum")
+    def merge_einsum(
+        self, x, batch_size, topk, seq_len, topk_values, topk_indices, one_hot
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:175]: jesteś w stanie `batch_size, topk, seq_len` wziąć z `one_hot`

File: research/conditional/moe_layers/expert_choice.py [hunk 9]
```diff
@@ -363,42 +151,47 @@ def extract_with_linear_bmm(
...
+            x,
+            topk_values,
+            one_hot,
+        )
+
+    def merge_bmm(
+        self, x, batch_size, topk, seq_len, topk_values, topk_indices, one_hot
+    ):
+        n_exp = self.n_experts
         _, topk, _ = topk_values.shape
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:191]: no właśnie tak, ale z `one_hot` miałbyś więcej

File: research/conditional/moe_layers/expert_choice.py [hunk 10]
```diff
@@ -410,92 +203,22 @@ def gating_postprocess_bmm(self, x, topk_values, one_hot, weight):
...
-            )
-        return x
-
-    def gating_postprocess_select(
+    def merge_index_select(
         self, x, batch_size, topk, seq_len, topk_values, topk_indices, one_hot
     ):
         # multiply by softmax
         with measure_time(self, "multiply_softmax"):
             ash.assert_shape("e k", topk_values, e=self.n_experts, k=topk)
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:219]: możemy to wywalić przy okazji?

File: research/conditional/moe_layers/expert_types.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        doutput: Optional[int] = None,
+    ):
+        super().__init__()
+        self.dmodel = dmodel
+        assert doutput is None or doutput == expert_size
+        self.doutput = expert_size
+        self.n_experts = n_experts
+
+        init = get_init_fun(init_type=init_type, init_scale=init_scale)
+        self.lin1_weight = init(shape=(n_experts, dmodel, expert_size), fan_in=dmodel)
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_types.py:148]: Can we make super class for common `init`?

**Review by Krystian-Krol — CHANGES_REQUESTED** [review]
Przejrzałem 13/21 plików - resztę dokończę jutro, póki co uwagi w komentarzach + wiele z nich się aplikuje bardziej globalnie ale w szczególności:
- Weź zdecyduj się co znaczy `topk`

File: research/conditional/moe_layers/expert_types.py (hunk 11 above)
Comment (ludziej) [reply research/conditional/moe_layers/expert_types.py:148]: yes, I've literally skipped this only because of how _some_ hate Inherentance, I'll gladly change that

File: research/conditional/moe_layers/expert_choice.py (hunk 10 above)
Comment (ludziej) [reply research/conditional/moe_layers/expert_choice.py:219]: niee, index select jest optymalny przy dużych modelach i dużych Granularity, ta implementacja jakbyśmy trenowali coś dużego jest domyślna.

File: research/conditional/moe_layers/expert_choice.py (hunk 9 above)
Comment (ludziej) [reply research/conditional/moe_layers/expert_choice.py:191]: API tych funkcji zakłada że nie znamy batch size i seq len, bo przy index select i grouping możemy ich nie znac i wtedy również nie ma one-hot.

File: research/conditional/moe_layers/expert_choice.py (hunk 8 above)
Comment (ludziej) [reply research/conditional/moe_layers/expert_choice.py:175]: to samo

File: research/conditional/tests/test_token_choice.py [hunk 12]
```diff
@@ -357,8 +364,8 @@ def make_expert():
 
         expert1 = make_expert()
         expert2 = make_expert()
-        expert_logic = ExpertRelu(dm, experts, exp_size, "kaiming_uniform", 1.0)
-        token_choice_layer = TokenChoiceFF(
+        expert_logic = ExpertFF(dm, experts, exp_size, "kaiming_uniform", 1.0)
```
[the remark this answers is not in the document]
Comment (ludziej) [reply research/conditional/tests/test_token_choice.py:367]: prawda, zmienię

File: research/conditional/utils/model_utils.py [hunk 13]
```diff
@@ -570,11 +674,11 @@ def get_mamba_layer(args):
     ]:
 
         def get_token_choice_ff(d_input, d_output):
-            return TokenChoiceFF(
+            return TokenChoiceFFOld(
                 dmodel=d_input,
                 doutput=d_output,
                 n_experts=args.n_experts,
-                expert_inner_function=ExpertRelu(
+                expert_inner_function=ExpertReluOld(
```
[the remark this answers is not in the document]
Comment (ludziej) [reply research/conditional/utils/model_utils.py:681]: powinno być old, w tamtym teście też powinno więc tam zmienię

File: research/conditional/moe_layers/expert_types.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+                self.lin2_weight,
+            )
+        else:
+            experts_output = torch.matmul(experts_output, self.lin2_weight)
+
+        assert experts_output.shape == (n_experts, capacity, self.doutput)
+        return experts_output
+
+
+class ExpertGated(ExpertFF):
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_types.py:72]: shouldn't it be `ExpertLinear`? and then `ExpertFF` also inherit from `ExpertLinear`?

File: research/conditional/moe_layers/expert_types.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        dmodel: int,
+        n_experts: int,
+        expert_size: int,
+        init_type: str,
+        init_scale: float,
+        use_einsum: bool = False,
+        doutput: Optional[int] = None,
+        activation_name: str = "silu",
+        topk: int = 1,
+        use_topk_initialization: bool = False,
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_types.py:84]: == `**kwargs`

File: research/conditional/moe_layers/moe_gating.py [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+                )
+        # each expert chooses k within dimension 1
+        if not self.group_by_batch and not self.softmax_ungrouped:
+            gate_out = gate_out.reshape(self.n_experts, batch_size * seq_len)
+        # perform softmax either over tokens for each expert or over experts for each token
+        with measure_time(self, "softmax"):
+            if self.softmax_over == "tokens":
+                gate_out = torch.softmax(gate_out, dim=1)
+            elif self.softmax_over == "experts":
+                gate_out = torch.softmax(gate_out, dim=0)
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/moe_gating.py:56]: what if we do typo in `softmax_over`? I believe this assert should be moved/ copied from `token_choice`/`expert_choice` responsibility into here

File: research/conditional/moe_layers/expert_choice.py [hunk 17]
```diff
@@ -205,65 +42,41 @@ def __init__(
...
-                (dmodel, n_experts),
-                fan_in=dmodel,
-                init_type=init_type,
-                scale=init_scale,
-            )
-        ).requires_grad_(True)
-        self.ln = LayerNorm(self.doutput) if use_layer_norm else None
+
+        init = get_init_fun(init_type=init_type, init_scale=init_scale)
+        gate = init((dmodel, n_experts), dmodel)
```
Comment (Krystian-Krol) [inline research/conditional/moe_layers/expert_choice.py:64]: dlaczego `expert_choice` definiuje gate?

File: research/conditional/tests/test_expert_choice.py [hunk 18]
```diff
@@ -127,29 +155,30 @@ def test_onehot_full_equivalence(self):
...
             exp_size,
             topk_fraction,
-            init_type="kaiming_uniform",
-            init_scale=1.0,
             one_hot_impl=True,
             group_by_batch=True,
+            use_torch_bmm=True,
         )
-        layer_einsum = ExpertChoiceFF(
+        layer_einsum = create_expert_choice(
```
Comment (Krystian-Krol) [inline research/conditional/tests/test_expert_choice.py:168]: `ExpertChoiceFFOld`?

File: research/conditional/tests/test_expert_choice.py [hunk 19]
```diff
@@ -169,16 +198,19 @@ def test_onehot_equivalence(self):
...
             exp_size,
             topk_fraction,
             init_type="kaiming_uniform",
             init_scale=1.0,
+            one_hot_impl=False,
             group_by_batch=True,
         )
-        layer_onehot = ExpertChoiceFF(
+
+        layer_onehot = create_expert_choice(
```
Comment (Krystian-Krol) [inline research/conditional/tests/test_expert_choice.py:213]: ?

File: research/conditional/utils/argparse.py [hunk 20]
```diff
@@ -246,6 +251,30 @@ def introduce_parser_arguments(
...
+        "--double_routing_before_relu",
+        action="store_true",
+        help="Whether second routing is calculated before relu",
+    )
+    parser.add_argument(
+        "--double_routing_snd_ln",
+        action="store_true",
+        help="Whether second routing has another layer norm",
+    )
+
```
Comment (Krystian-Krol) [inline research/conditional/utils/argparse.py:277]: is it used anywhere yet?

File: research/conditional/utils/model_utils.py [hunk 21]
```diff
@@ -336,16 +354,57 @@ def get_expert_choice_args(args):
...
 
-        experts_per_token = args.effective_dff / expert_size
-
-        topk_fraction = experts_per_token / args.n_experts
-        assert 0.0 <= topk_fraction <= 1.0
-        args.topk_fraction = topk_fraction
+        args.routing_top_k = args.effective_dff / expert_size
+        args.topk_fraction = args.routing_top_k / args.n_experts
+        assert 0.0 <= args.topk_fraction <= 1.0
     else:
```
Comment (Krystian-Krol) [inline research/conditional/utils/model_utils.py:360]: a co jeśli ustawiliśmy `set_arguments_option3`?

**Review by Krystian-Krol — COMMENTED** [review]
Jeszcze drobne komentarze, w szczególności początek `model_utils` musimy przegadać na żywo

File: research/conditional/utils/argparse.py (hunk 20 above)
Comment (ludziej) [reply research/conditional/utils/argparse.py:277]: i'll remove it

**Review by Krystian-Krol — APPROVED** [review]
GG, LETS GO
